### PR TITLE
some notes on transaction and block security

### PR DESCRIPTION
**Pull request by hxw**
Signed-off-by: Christopher Hall <hsw@bitmark.com>

### CODE REVIEW

File: bitmark-references/security.markdown [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+## Block Identifiers
+
+The Block Identifiers are used both to link the previous block and as
+the Proof of Work for the block chain.  Since SHA3 is a fast algorithm
+that can be performed on minimal hardware it was decided to use a
+memory hard algorithm for the Block Identifiers to prevent the use of
+ASIC miners and to set the memory parameters to make GPUs uneconomic
+leaving CPUs as the best way to perform the Proof of Work.  This
+prevents existing Bitcoin and Litecoin mining hardware from using
+their large system again the chain.
```
Comment (araujobsd) [inline bitmark-references/security.markdown]: s/again/against/g

File: bitmark-references/security.markdown [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+So the Argon2 was chosen as a modern high security memory hard hashing
+algorithm from: https://github.com/P-H-C/phc-winner-argon2
+
+The Block Header is made from the Argon2 Hash of the previous block,
+the root Merkle hash covering all transactions and the number of the
+block.  Other fields cover timestamp, difficulty and NONCE.  The
+difficulty is scaled differently from Bitcoin and allows for much
+lower hashing rates; while a modern CPU can run SHA2/SHA3 at many
+thousands of hashes per second, it con only perform one or two Argon2
```
Comment (araujobsd) [inline bitmark-references/security.markdown]: s/con/can/g

**Review by araujobsd — APPROVED** [review]
There are 2 misspellings, overall looks good to me.

File: bitmark-references/security.markdown (hunk 2 above)
Comment (araujobsd) [inline bitmark-references/security.markdown]: s/it con/it can/g

**Review by araujobsd — APPROVED** [review]

File: bitmark-references/security.markdown [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+hashes per second because of the increase complexity of the Algorithm.
+The NONCE is used for the hashing program to adjust to find a has that
```
Comment (nodestory) [inline bitmark-references/security.markdown]:
```suggestion
The NONCE is used for the hashing program to adjust to find <what?> that
```

File: bitmark-references/security.markdown [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+The reason for signing over the signature bytes is to protect against
+signature substitution.
+
+The Transaction Identifiers are merged into a single root hash using
+the Merkle Tree Algorithm: https://brilliant.org/wiki/merkle-tree/
+
+Since the tree is sensitive to missing or changed hashes and to the
+order of the hashes and produces a single result it can be used to
+minimise the amount of data that is stored in the block header to
+secure the transactions.  The Bitmark Blockchain used SHA3-256 for the
```
Comment (nodestory) [inline bitmark-references/security.markdown]: Why does `sensitive to the order of hashes` lead to `generate a single result to save disk space`?

File: bitmark-references/security.markdown [hunk 5]
```diff
@@ -0,0 +1,73 @@
+# Security feature of the Bitmark Blockchain
+
```
Comment (shannona) [inline bitmark-references/security.markdown:2]: This is another article where we really need a good intro that says what the article is about.

Something like:

"The Bitmark Property System uses signatures and hashes to secure its transactions and blocks."

File: bitmark-references/security.markdown [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+## Transactions
+
```
Comment (shannona) [inline bitmark-references/security.markdown:7]: For the sections, we really need to say how each section relates to the overall topic (of security).

For this section, we might replace the first three lines with something like:

Individual transactions are secured by cryptographic signatures. Single signatures are used for basic transactions, while countersigned signatures are used for special transfers.

File: bitmark-references/security.markdown [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+The Ed25519 signature algorithm details can be viewed at: https://ed25519.cr.yp.to/
+
+The important features for the Bitmark Block chain are:
+* high security signatures with collision resistance
+* small 32 byte keys
+* signature is only 64 bytes
+* fast verification
+
+## Transaction Identifiers
+
```
Comment (shannona) [inline bitmark-references/security.markdown:32]: Again, we want to stick with the topic of security.

Something like: "Transactions are securely incorporated into a block through the SHA3-256 algorithm, which is used to derive a 32-byte identifier for each transaction." ...

File: bitmark-references/security.markdown [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+the Merkle Tree Algorithm: https://brilliant.org/wiki/merkle-tree/
+
+Since the tree is sensitive to missing or changed hashes and to the
+order of the hashes and produces a single result it can be used to
+minimise the amount of data that is stored in the block header to
+secure the transactions.  The Bitmark Blockchain used SHA3-256 for the
+Merkle Tree generation.
+
+## Block Identifiers
+
```
Comment (shannona) [inline bitmark-references/security.markdown:62]: And once more, we need to push on security to tie everything together. 

Something like:

The SHA3 algorithm also creates security for the blocks themselves, though the Block Identifiers. ...

**Review by shannona — COMMENTED** [review]
This document needs to be revised to focus on the central topic of "security". I've offered some suggestions for doing so, but think overall it could benefit from a review with that consideration in mind.

**Comment by vienlt** [comment]
@hxw: Could you check the comments and update the document if you feel reasonable? And please check and fix to make the commit pass the CI (I think it would require to merge the master to your branch).

**Comment by hxw** [comment]
I checked off all the outstanding comments and added a few more fixes

**Comment by vienlt** [comment]
Merge for the first deployment.

**Pull request merged**